src/target/riscv/riscv.c

```diff
@@ -5782,6 +5861,19 @@ int riscv_enumerate_triggers(struct target *target)
 		return ERROR_OK;
 	}
 
+	/* Obtaining tinfo.version value once.
+	 * No need to enumerate per-trigger.
+	 * See https://github.com/riscv/riscv-debug-spec/pull/1081.
+	 */
+	int tinfo_version;
+	riscv_reg_t tinfo;
+	if (riscv_reg_get(target, &tinfo, GDB_REGNO_TINFO) == ERROR_OK)
+		tinfo_version = get_field(tinfo, CSR_TINFO_VERSION);
+	else
+		tinfo_version = RISCV_TINFO_VERSION_UNKNOWN;
+	r->tinfo_version = tinfo_version;
+	LOG_TARGET_DEBUG(target, "Trigger tinfo.version = %" PRId32 ".", r->tinfo_version);
+
```

Comment:
Two small things:

- When printing a value of type `int`, use `%d`. `PRId32` is strictly intended for `uint32_t`. (https://en.cppreference.com/w/c/io/fprintf, https://en.cppreference.com/w/c/types/integer)
- Please say `Trigger tinfo.version is unknown` instead of `Trigger tinfo.version = -1`.

```suggestion
	riscv_reg_t tinfo;
	if (riscv_reg_get(target, &tinfo, GDB_REGNO_TINFO) == ERROR_OK) {
		r->tinfo_version = get_field(tinfo, CSR_TINFO_VERSION);
		LOG_TARGET_DEBUG(target, "Trigger tinfo.version = %d.", r->tinfo_version);		
	} else {
		tinfo_version = RISCV_TINFO_VERSION_UNKNOWN;
		LOG_TARGET_DEBUG(target, "Trigger tinfo.version is unknown.", r->tinfo_version);		
	}

```

Reply:
We're assuming a vaid trigger is selected before reading `Trigger tinfo` , is that assumption correct? @JanMatCodasip

Reply:
My interpretation of the debug spec is that `tinfo.version` is a hardwired constant, independent on the value of `tselect`.

So your approach looks good to me. I believe there is no need to write anyting to `tselect` before reading `tinfo.version`.

Reply:
addressed, thanks a lot @JanMatCodasip